File: blue-sdk-android/blue-sdk/src/main/kotlin/com/blue/sdk/transport/BLEScanner.kt
// BLEScanner.kt
// BlueSDK - BLE 设备扫描器（FR01）

package com.blue.sdk.transport

import android.bluetooth.BluetoothAdapter
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.os.Handler
import android.os.Looper
import com.blue.sdk.error.BlueError
import com.blue.sdk.internal.BlueLogger
import com.blue.sdk.internal.CallbackDispatcher
import com.blue.sdk.model.ScannedDevice

internal class BLEScanner {

    companion object {
        private const val DEVICE_NAME_PREFIX = "LX-PD02"
        private const val SCAN_DELAY_AFTER_DISCONNECT_MS = 500L
    }

    @Volatile internal var isScanning = false
    private var onDeviceFound: ((ScannedDevice) -> Unit)? = null
    private var onError: ((BlueError) -> Unit)? = null
    private val handler = Handler(Looper.getMainLooper())
    private var pendingScanRunnable: Runnable? = null

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val name = result.device.name ?: return
            if (!name.startsWith(DEVICE_NAME_PREFIX)) return
            val device = ScannedDevice(
                deviceId = result.device.address,
                deviceName = name,
                rssi = result.rssi
            )
            BlueLogger.debug("发现设备：$name，RSSI：${result.rssi}")
            CallbackDispatcher.dispatch { onDeviceFound?.invoke(device) }
        }

        override fun onScanFailed(errorCode: Int) {
            BlueLogger.error("BLE 扫描失败，错误码：$errorCode")
            CallbackDispatcher.dispatch { onError?.invoke(BlueError.BleError(Exception("Scan failed: $errorCode"))) }
        }
    }

    fun startScan(
        adapter: BluetoothAdapter,
        onDeviceFound: (ScannedDevice) -> Unit,
        onError: (BlueError) -> Unit
    ) {
        pendingScanRunnable?.let { handler.removeCallbacks(it) }
        pendingScanRunnable = null

        if (isScanning) {
            adapter.bluetoothLeScanner?.stopScan(scanCallback)
            isScanning = false
        }

        this.onDeviceFound = onDeviceFound
        this.onError = onError

        val runnable = Runnable {
            isScanning = true
            val settings = ScanSettings.Builder()
                .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
                .build()
            adapter.bluetoothLeScanner?.startScan(null, settings, scanCallback)
            BlueLogger.info("BLE 扫描已启动，过滤前缀：$DEVICE_NAME_PREFIX")
        }
        pendingScanRunnable = runnable
        handler.postDelayed(runnable, SCAN_DELAY_AFTER_DISCONNECT_MS)
    }

    fun stopScan(adapter: BluetoothAdapter) {
        pendingScanRunnable?.let { handler.removeCallbacks(it) }
        pendingScanRunnable = null

        if (!isScanning) return
        adapter.bluetoothLeScanner?.stopScan(scanCallback)
        isScanning = false
        onDeviceFound = null
        onError = null
        BlueLogger.info("BLE 扫描已停止")
    }
}

File: blue-sdk-ios/BlueSDK/BlueSDK/Classes/Transport/BLEScanner.swift
// BLEScanner.swift
// BlueSDK - LX-PD02 智能药盒蓝牙通信 SDK
//
// BLE 设备扫描器：过滤广播名前缀为 LX-PD02 的设备（FR01）
// 使用 BLECentralManager 单例，确保与连接器共用同一个 CBCentralManager

import Foundation
import CoreBluetooth

/// 扫描到的设备信息
@objc public class ScannedDevice: NSObject {
    /// 设备唯一标识（UUID 字符串）
    @objc public let deviceId: String
    /// 设备广播名称（如 LX-PD02-A1B2）
    @objc public let deviceName: String
    /// 信号强度（RSSI，单位 dBm）
    @objc public let rssi: Int
    /// 底层 CBPeripheral（内部使用）
    internal let peripheral: CBPeripheral

    internal init(peripheral: CBPeripheral, rssi: Int) {
        self.deviceId = peripheral.identifier.uuidString
        self.deviceName = peripheral.name ?? "Unknown"
        self.rssi = rssi
        self.peripheral = peripheral
    }
}

/// BLE 扫描器代理协议（供 BLECentralManager 回调转发）
protocol BLEScannerDelegate: AnyObject {
    func bleCentralManagerDidUpdateState(_ central: CBCentralManager)
    func bleCentralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber)
}

/// BLE 设备扫描器
final class BLEScanner: NSObject, BLEScannerDelegate {

    // MARK: - 常量

    private static let deviceNamePrefix = "LX-PD02"
    private static let scanDelayAfterDisconnect: TimeInterval = 0.5

    // MARK: - 状态

    private(set) var isScanning = false
    private var onDeviceFound: ((ScannedDevice) -> Void)?
    private var onError: ((BlueError) -> Void)?
    private let logger = BlueLogger.shared
    private let centralManager = BLECentralManager.shared
    private var pendingScanTimer: Timer?

    // MARK: - 公开方法

    /// 开始扫描 LX-PD02 设备
    /// - Parameters:
    ///   - onDeviceFound: 发现设备回调（主线程）
    ///   - onError: 错误回调（主线程）
    func startScan(
        onDeviceFound: @escaping (ScannedDevice) -> Void,
        onError: @escaping (BlueError) -> Void
    ) {
        pendingScanTimer?.invalidate()
        pendingScanTimer = nil

        if isScanning {
            centralManager.centralManager.stopScan()
            isScanning = false
        }

        self.onDeviceFound = onDeviceFound
        self.onError = onError

        centralManager.scanDelegate = self

        if centralManager.state == .poweredOn {
            pendingScanTimer = Timer.scheduledTimer(withTimeInterval: BLEScanner.scanDelayAfterDisconnect, repeats: false) { [weak self] _ in
                self?.startScanImmediately()
            }
        } else {
            logger.debug("Bluetooth not ready, waiting for state change")
        }
    }

    private func startScanImmediately() {
        guard !isScanning else { return }
        isScanning = true
        centralManager.centralManager.scanForPeripherals(withServices: nil, options: [
            CBCentralManagerScanOptionAllowDuplicatesKey: false
        ])
        logger.info("BLE scan started, prefix filter: \(BLEScanner.deviceNamePrefix)")
    }

    /// 停止扫描
    func stopScan() {
        pendingScanTimer?.invalidate()
        pendingScanTimer = nil

        guard isScanning else { return }
        centralManager.centralManager.stopScan()
        isScanning = false
        onDeviceFound = nil
        onError = nil
        logger.info("BLE scan stopped")
    }

    // MARK: - 私有方法

    private func startScanIfReady() {
        guard centralManager.state == .poweredOn, onDeviceFound != nil else { return }
        startScanImmediately()
    }

    // MARK: - BLEScannerDelegate

    func bleCentralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn:
            logger.info("Bluetooth powered on")
            startScanIfReady()
        case .poweredOff:
            logger.warn("Bluetooth powered off")
            CallbackDispatcher.shared.dispatch { [weak self] in
                self?.onError?(.bleError(underlying: nil))
            }
        case .unauthorized:
            logger.warn("Bluetooth permission denied")
            CallbackDispatcher.shared.dispatch { [weak self] in
                self?.onError?(.permissionDenied)
            }
        default:
            break
        }
    }

    func bleCentralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber) {
        // 过滤广播名前缀
        guard let name = peripheral.name,
              name.hasPrefix(BLEScanner.deviceNamePrefix) else { return }

        let device = ScannedDevice(peripheral: peripheral, rssi: RSSI.intValue)
        logger.debug("Device found: \(name), RSSI: \(RSSI)")

        CallbackDispatcher.shared.dispatch { [weak self] in
            self?.onDeviceFound?(device)
        }
    }
}
